packages/dapi/test/unit/grpcServer/handlers/blockheaders-stream/subscribeToBlockHeadersWithChainLocksHandlerFactory.spec.js

```diff
@@ -0,0 +1,166 @@
+
+  beforeEach(function () {
+    coreAPIMock = {
+      getBlock: this.sinon.stub(),
+      getBestBlockHeight: this.sinon.stub(),
+      getBlockHash: this.sinon.stub(),
+      getBestChainLock: this.sinon.stub(),
+    };
+    subscribeToNewBlockHeadersMock = this.sinon.stub();
+
+    async function* asyncGenerator() {
+      yield [{ toBuffer: () => Buffer.from('fake', 'utf-8') }];
+    }
+
+    getHistoricalBlockHeadersIteratorMock = () => asyncGenerator();
+    zmqClientMock = { on: this.sinon.stub(), topics: { hashblock: 'fake' } };
+
+    // eslint-disable-next-line operator-linebreak
+    subscribeToBlockHeadersWithChainLocksHandler =
+      subscribeToBlockHeadersWithChainLocksHandlerFactory(
+        getHistoricalBlockHeadersIteratorMock,
+        coreAPIMock,
+        zmqClientMock,
+        subscribeToNewBlockHeadersMock,
+      );
+  });
+
+  it('should subscribe to newBlockHeaders', async function () {
+    this.sinon.stub(AcknowledgingWritable.prototype, 'write');
+    const request = new BlockHeadersWithChainLocksRequest();
+
+    call = new GrpcCallMock(this.sinon, request);
+
+    call.request.setFromBlockHash('fakehash');
+    call.request.setCount(0);
+
+    coreAPIMock.getBestChainLock.resolves({
+      height: 1,
+      signature: Buffer.from('fakeSig'),
+      blockHash: Buffer.from('fakeHash'),
+    });
+    coreAPIMock.getBlock.resolves({ height: 1 });
+
+    await subscribeToBlockHeadersWithChainLocksHandler(call);
+    expect(subscribeToNewBlockHeadersMock).to.have.been.called();
+  });
+
+  it('should subscribe from block hash', async function () {
+    const writableStub = this.sinon.stub(AcknowledgingWritable.prototype, 'write');
+    const request = new BlockHeadersWithChainLocksRequest();
+
+    call = new GrpcCallMock(this.sinon, request);
+
+    call.request.setFromBlockHash('someBlockHash');
+    call.request.setCount(0);
+
+    coreAPIMock.getBestChainLock.resolves({
+      height: 1,
+      signature: Buffer.from('fakesig', 'hex'),
+      blockHash: Buffer.from('fakeHash', 'hex'),
+    });
+
+    try {
+      coreAPIMock.getBlock.resolves({ height: -1 });
+      await subscribeToBlockHeadersWithChainLocksHandler(call);
+    } catch (e) {
+      expect(e).to.be.instanceOf(NotFoundGrpcError);
+      expect(e.message).to.be('fromBlockHash is not found');
+    }
+
+    try {
+      coreAPIMock.getBlock.resolves({ height: 1, confirmations: -1 });
+      await subscribeToBlockHeadersWithChainLocksHandler(call);
+    } catch (e) {
+      expect(e).to.be.instanceOf(NotFoundGrpcError);
+      expect(e.message.includes('is not part of the best block chain')).to.be.true();
+    }
+
+    try {
+      coreAPIMock.getBlock.resolves({ height: 10 });
+      coreAPIMock.getBestBlockHeight.resolves(11);
+      await subscribeToBlockHeadersWithChainLocksHandler(call);
+    } catch (e) {
+      expect(e).to.be.instanceOf(InvalidArgumentGrpcError);
+    }
+
+    await subscribeToBlockHeadersWithChainLocksHandler(call);
```

Comment:
Shouldn't we move all these cases to separate tests

Reply:
Moved error handling to a separate test case